src/metabase/lib/remove_replace.cljc

```diff
@@ -551,7 +567,16 @@
      (remove-join query stage-number join-spec)
      (update-joins query stage-number join-spec (fn [joins join-alias]
                                                   (mapv #(if (= (:alias %) join-alias)
-                                                           new-join
+                                                           (cond-> new-join
+                                                             (or (not= (:conditions new-join)
+                                                                       (:conditions %))
+                                                                 (not= (:source-table new-join)
+                                                                       (:source-table %))
+                                                                 (not= (:source-card new-join)
+                                                                       (:source-card %)))
```

Comment:
This makes me nervous. I'd rather see `(not= (dissoc new-join :some :bits) (dissoc % :some :bits))` so we're not vulnerable to a new field getting added which matters but isn't mentioned here. (That feels slightly riskier than the reverse problem, of removing the alias sometimes when it didn't need to be.)